pkg/new-ui/v1beta1/frontend/package.json

```diff
@@ -0,0 +1,69 @@
+{
+  "name": "frontend",
+  "version": "0.0.0",
+  "scripts": {
+    "ng": "ng",
+    "format": "prettier --write './**/*.{ts,html,css}'",
+    "start": "npm run copyLibAssets && ng serve --proxy-config proxy.conf.json",
```

Comment:
Correct, it is used only for local testing to have the frontend and the backend [ either run directly or kubectl-port-forwarded ] under the same origin

Reply:
Do we want to have this proxy for the local testing since we explain how to run frontend and backend (user should execute `go run main.go`) in the README: https://github.com/arrikto/katib/tree/feature-kimwnasptd-katib-ui/pkg/ui/v1beta1#serve-ui-frontend-and-backend.
I assume `npm run start` is used only to serve frontend ?

Reply:
> I assume npm run start is used only to serve frontend ?

Correct, it creates a local webpack dev server that watches the source code and reloads the server on changes.

> Do we want to have this proxy for the local testing since we explain how to run frontend and backend (user should execute go run main.go) in the README: https://github.com/arrikto/katib/tree/feature-kimwnasptd-katib-ui/pkg/ui/v1beta1#serve-ui-frontend-and-backend.

I think it makes sense to also have instructions on how to only run the frontend, without the backend. There might be cases where someone would like to use a `kubectl port-forward` to and use the backend from a cluster [this is actually mostly what I do]. This is why I'd, personally, like to keep the proxy and run with `npm run start`.

But I don't have an issue on providing instructions in the README  only on how to serve the frontend via the backend. But lets also keep this proxy file

Reply:
In that case, can you please update README instructions how to:
1. Serve only frontend, without the proxy.
2. Serve frontend with port-forwarding backend (I believe it is current `npm run start` script)
3. Serve frontend and backend by running `go run main.go` (which is already explaining here: https://github.com/arrikto/katib/tree/feature-kimwnasptd-katib-ui/pkg/new-ui/v1beta1#serve-ui-frontend-and-backend)

Reply:
Sure, but
> Serve only frontend, without the proxy.
Serving the frontend files alone, via a webpack dev server, doesn't make much sense without proxying. The dev server will be exposing a different port than the backend's port, which makes the frontend unable to make requests to backend due to CORS.

But I will update the README to include steps 2 and 3

Reply:
> Serving the frontend files alone, via a webpack dev server, doesn't make much sense without proxying. The dev server will be exposing a different port than the backend's port, which makes the frontend unable to make requests to backend due to CORS.

We still can debug frontend without backend running, right ?
For example, to fix some styling or forms.
Do you think it is redundant ?

Reply:
> We still can debug frontend without backend running, right ?

Correct. 

> For example, to fix some styling or forms.
Do you think it is redundant ?

No it's not redundant, this makes total sense. But without a running backend we won't be able to fetch data, so the things we could see will be limited.

Reply:
I've pushed another commit to update the README, and introduce a new npm script for building the UI in watch mode.
PTAL

Reply:
> But without a running backend we won't be able to fetch data, so the things we could see will be limited.

We can manually add data to the appropriate variables since we know the post responses.

>introduce a new npm script for building the UI in watch mode.
 
Thank you! Can you also explain what `npm run build:watch` script is doing in the README, please ?

Reply:
I tried to briefly explain what it does in the README, but in just a sentence https://github.com/kubeflow/katib/pull/1427/commits/bc676bdca76ce64b807d7cb0e562875519d81bba#diff-d84e6861c045b786e7203bb6b0f7334f42049f4da6b316636dfc9d4b12b319b5R73

I think that if a user would like to understand the more technical details of how the dev server starts and constantly watches the source files then they could take a look at the `package.json` script or the Angular docs https://angular.io/cli/build.

Would you like me to further explain the `build:watch` script in the README?

Reply:
I left comment https://github.com/kubeflow/katib/pull/1427#discussion_r579521116.

I was thinking maybe we can leave only 2 scripts for the build:
`npm run build:prod` and `npm run build:watch` with consistent flags and differences only in `--watch` and `--prod`` flag.
For instance:

```
package.json:

"build:watch": "npm run copyLibAssets && ng build --base-href /katib/ --deploy-url /katib/static/  --outputPath dist/default --watch",
"build:prod": "npm run copyLibAssets && ng build --base-href /katib/ --deploy-url /katib/static/ --outputPath dist/default --prod",
```

It makes very clear for the user what is the differences in the scripts and we don't need to modify outputPath in the Dockerfile:
https://github.com/kubeflow/katib/blob/bc676bdca76ce64b807d7cb0e562875519d81bba/cmd/new-ui/v1beta1/Dockerfile#L33

Reply:
Got you, I think we can do both this and fix your other comment about the two COPYs in the Dockerfile https://github.com/kubeflow/katib/pull/1427#discussion_r579518359

Yes, lets leave only two scripts for build as you mentioned but we will need to change the outputPath in both cases to be `--outputPath dist/static`. This is even the case for the current Dockerfile, where I was copying the files from `dist/default` to static. So there's no reason for the build folder to not be `static` from the beginning
```dockerfile
COPY --from=frontend /src/dist/default /app/build/static/
```

So making the scripts look similar, with the outputPath change into consideration, will both make the npm scripts easier to see/understand and also simplify a bit the dockerfile. Fixing it right away!

Reply:
Make sense, thank you @kimwnasptd!

Reply:
Pushed a commit that does what we describe above.
I also tested that serving the UI works both locally and after building the updated container image